stateproof/builder.go

```diff
@@ -367,36 +388,30 @@ func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.Forw
 		spw.log.Infof("spw.handleSig: starts gathering signatures for round %d", sfa.Round)
 	}
 
-	pos, ok := builderForRound.AddrToPos[sfa.SignerAddress]
-	if !ok {
-		return network.Disconnect, fmt.Errorf("handleSig: %v not in participants for %d", sfa.SignerAddress, sfa.Round)
+	sig := pendingSig{
+		signer:       sfa.SignerAddress,
+		sig:          sfa.Sig,
+		fromThisNode: sender == nil,
 	}
-
-	if isPresent, err := builderForRound.Present(pos); err != nil || isPresent {
-		// Signature already part of the builderForRound, ignore.
+	err := builderForRound.insertSig(&sig, true)
+	if errors.Is(err, errSigAlreadyPresentAtPos) {
+		// Safe to ignore this error as it means we already have a valid signature for this address
 		return network.Ignore, nil
 	}
-
-	if err := builderForRound.IsValid(pos, &sfa.Sig, true); err != nil {
+	if errors.Is(err, errFailedToAddSigAtPos) {
+		return network.Ignore, err
+	}
+	if errors.Is(err, errAddressNotInVoters) || errors.Is(err, errSignatureVerification) {
 		return network.Disconnect, err
 	}
 
```

Comment:
Should probably positively verify that err is nil here, to avoid future bugs involving modification of insertSig.